# Importing libraries

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os 

In [2]:
path = r'C:\Users\Dan\Desktop\Career Foundry\Instacart Basket Analysis'

In [3]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [5]:
# create a data frame
df_test = pd.DataFrame()

In [6]:
# create a mixed column type
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
df_test['mix'] = df_test['mix'].astype('str')

changed to string

In [11]:
# check for missing data
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
#create subset of column containing nulls
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
df_prods.shape

(49693, 5)

In [16]:
df_nan.shape

(16, 5)

In [17]:
#create new dataframe without omitting nulls
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [18]:
df_prods_clean.shape

(49677, 5)

In [19]:
#search for duplicate rows
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [22]:
df_prods_clean.shape

(49677, 5)

In [23]:
#create new date frame that doesnt include duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [24]:
df_prods_clean_no_dups.shape

(49672, 5)

# Exercise

# Q2

In [26]:
df_prods . describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [27]:
df_ords . describe ()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

The max price value in df_prods seems extremely high ($99,999). This is surely an error as we are looking at supermarket prices. 

# Q3

In [32]:
# checking for mixed data types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col, ' mixed')
  else: print (col, ' consistent')

Unnamed: 0  consistent
order_id  consistent
user_id  consistent
eval_set  consistent
order_number  consistent
orders_day_of_week  consistent
order_hour_of_day  consistent
days_since_last_order  consistent


no mixed datatypes found

# Q4

no mixed datatypes found

# Q5

In [34]:
df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
eval_set                      0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

checked for missing values

In [35]:
df_ords[df_ords['user_id'] == 1]

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0            0   2539329        1    prior             1                   2   
1            1   2398795        1    prior             2                   3   
2            2    473747        1    prior             3                   3   
3            3   2254736        1    prior             4                   4   
4            4    431534        1    prior             5                   4   
5            5   3367565        1    prior             6                   2   
6            6    550135        1    prior             7                   1   
7            7   3108588        1    prior             8                   1   
8            8   2295261        1    prior             9                   1   
9            9   2550362        1    prior            10                   4   
10          10   1187899        1    train            11                   4   

    order_hour_of_day  days_since_last_order  
0                   8                    NaN  
1                   7                   15.0  
2                  12                   21.0  
3                   7                   29.0  
4                  15                   28.0  
5                   7                   19.0  
6                   9                   20.0  
7                  14                   14.0  
8                  16                    0.0  
9                   8                   30.0  
10                  8                   14.0

In [36]:
df_ords[df_ords['user_id'] == 2]

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
11          11   2168274        2    prior             1                   2   
12          12   1501582        2    prior             2                   5   
13          13   1901567        2    prior             3                   1   
14          14    738281        2    prior             4                   2   
15          15   1673511        2    prior             5                   3   
16          16   1199898        2    prior             6                   2   
17          17   3194192        2    prior             7                   2   
18          18    788338        2    prior             8                   1   
19          19   1718559        2    prior             9                   2   
20          20   1447487        2    prior            10                   1   
21          21   1402090        2    prior            11                   1   
22          22   3186735        2    prior            12                   1   
23          23   3268552        2    prior            13                   4   
24          24    839880        2    prior            14                   3   
25          25   1492625        2    train            15                   1   

    order_hour_of_day  days_since_last_order  
11                 11                    NaN  
12                 10                   10.0  
13                 10                    3.0  
14                 10                    8.0  
15                 11                    8.0  
16                  9                   13.0  
17                 12                   14.0  
18                 15                   27.0  
19                  9                    8.0  
20                 11                    6.0  
21                 10                   30.0  
22                  9                   28.0  
23                 11                   30.0  
24                 10                   13.0  
25                 11                   30.0

looking in to user_id and when NaN values appear

In [41]:
#Creating subset for missing values

df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

created subset

In [43]:
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_last_order  
0                         2                  8                    NaN  
11                        2                 11                    NaN  
26                        1                 14                    NaN  
39                        6                 11                    NaN  
45                        3                 12                    NaN  
...                     ...                ...                    ...  
3420930                   4                 12                    NaN  
3420934                   3                 18                    NaN  
3421002                   6                 19                    NaN  
3421019                   1                 15                    NaN  
3421069                   3                 11                    NaN  

[206209 rows x 8 columns]

the reason for all the missing values is that each time a new customer places their first order the days_since_last_order isn't applicable therefore it is left blank. 

# Q6

I have addressed the missing values by deciding to leave them as is. This is because the rows still contain lots of valuable data and removing 206209 rows would mean a lot of data lost. 

# Q7

In [44]:
#search for duplicate rows
df_ords_dups = df_ords[df_ords.duplicated()]

In [45]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

In [46]:
df_ords_dups.size

0

no duplicates found

# Q8

no duplicates found

# Q9
Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [48]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [49]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))